Add note comments to enforce nondeterministic alert documentation

kurtamohler opened the pull request:
This PR fulfills Ed's request (https://github.com/pytorch/pytorch/pull/41692#discussion_r473122076) for a strategy to keep the functions that have nondeterministic alerts fully documented.

Part of #15359

kurtamohler committed:
Add note comments to enforce nondeterministic alert documentation

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,37 @@ class CAFFE2_API Context {
   bool deterministic() const;
   void setDeterministic(bool);
+
+  // This method should be called by operations that act nondeterministically, so that they
```
Let's start by describing what the function actually does first, then show an example of how to use it. Maybe something like, "throws an a RuntimeError if `Context::deterministic()` is true." This should be called before performing a nondeterministic operation like this:

<example description>
<NOTE>
alertNotDeterministic(<caller>)
<nondeterministic operations here>

kurtamohler replied:
Done

ezyang commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -145,6 +145,9 @@ const char* _cublasGetErrorEnum(cublasStatus_t error) {
 void gemm<double>(CUDABLAS_GEMM_ARGTYPES(double)) {
+  // NOTE: If you copy and paste this alert to a different operation, you must
+  //    document the operation in torch/__init__.py:set_deterministic(), and copy
+  //    and paste this note.
```
I think you should try to make this just one line, to reduce the likelihood that we have to do a giant multi-line sed everywhere if we need to update this documentation. Maybe `// Before modifying/copying this code, please see Note [set_deterministic]`

kurtamohler replied:
Good idea. How about this, to make it a bit easier to find: `NOTE: Before modifying/copying, see aten/src/ATen/Contex.h:alertCuBLASConfigNotDeterministic`

And then for the other non-CuBLAS alerts: `NOTE: Before modifying/copying, see aten/src/ATen/Contex.h:alertNotDeterministic`

mruberry replied:
As usual Ed and I appear to have the same idea. Your idea sounds good but I think you can just refer to the note's name (if you like the note approach).

kurtamohler replied:
Ah, I see now that this "Note [...]" convention is used often throughout PyTorch, so I agree that this is the best way.

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
   bool deterministic() const;
   void setDeterministic(bool);
+
+  // Throws an error if `Context::deterministic()` is true. This method should be called by
```
How about promoting this to a proper note:

```
# Note [Writing Nondeterministic Operations]
```

That way the comments can clearly refer to this note for more detail.

The note can then start by explaining the three things nondeterministic operations must do:

- Write a comment explaining why the operation is nondeterministic
- Add the call to alertNotDeterministic and the comment explaining that all nondeterministic operations should make this call
- Update the documentation of set_deterministic() in torch/__init__.py

It can even include an example, if you like

mruberry replied:
(The note should probably include additional requirements, like when to call `alertCuBLASConfigNotDeterministic`, too.)

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
+  // All callsites should have a comment explaining why the operation is nondeterministic.
+  //
+  // To help enforce full documentation, all callsites should also have the following note:
+  // NOTE: If you copy and paste this alert to a different operation, you must add a
```
Would it be OK if this note was just a one-line? 

"See the note [Writing Nondeterministic Operations] when implementing a nondeterministic operation"?

That might make maintenance easier in the future.

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
+  // NOTE: If you copy and paste this alert to a different operation, you must add a
+  //    comment explaining why the operation is nondeterministic, document the operation
+  //    in torch/__init__.py:set_deterministic(), and copy and paste this note.
   void alertNotDeterministic(c10::string_view const& caller);
```
If the note appears first, the description of what this function does could go here, just above the function. "Throws an error when..."

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
   bool allowTF32CuBLAS() const;
   void setAllowTF32CuBLAS(bool);
+
+  // Throws an error if `Context::deterministic()` is true and CUDA >= 10.2 and
```
grammar nit: "is true, CUDA >= 10.2, and ..."

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
   void setAllowTF32CuBLAS(bool);
+
+  // Throws an error if `Context::deterministic()` is true and CUDA >= 10.2 and
+  // CUBLAS_WORKSPACE_CONFIG is not set deterministically. This method should be called
```
What does it mean for CUBLAS_WORKSPACE_CONFIG to "[not be] set deterministically"?

kurtamohler replied:
I've changed this to describe exactly which two settings won't cause an error.

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
+  // Throws an error if `Context::deterministic()` is true and CUDA >= 10.2 and
+  // CUBLAS_WORKSPACE_CONFIG is not set deterministically. This method should be called
+  // by operations that can act nondeterministically if the CuBLAS workspace setting is
+  // not set properly. For more details:
```
Is the setting actually improper or does it just allow for nondeterminism?

mruberry commented on aten/src/ATen/Context.h:
```diff
@@ -114,12 +114,42 @@ class CAFFE2_API Context {
+  // Most of the time, this method should be called like this:
+  //  at::globalContext().alertCuBLASConfigNotDeterministic();
+  //
+  // To help enforce full documentation, all callsites should have the following note:
```
See above comments. Maybe most of the description for these two functions can be combined in the single rollup note so people don't read the first notice and fail to see this second one?

kurtamohler commented:
It occurred to me that `Context::deterministic()` should probably have a similar note, since we want all operations that use it to be documented as well. I think we can combine that into this PR.

kurtamohler commented:
I believe I've addressed everything. I've added the `Note [Writing Nondeterministic Operations]`, as well as `Note [Enabling Deterministic Operations]`, because I think both are probably important for enforcing good documentation.

mruberry reviewed (approved):
Looks great! @ezyang, would you like to look again?

ezyang reviewed (approved):
Thanks, much appreciated
